### PR TITLE
[BUG] Fix a bug in PeriodIndex arithmetic

### PR DESCRIPTION
**Pull request by khrapovs**
#### Reference Issues/PRs

Fixes https://github.com/alan-turing-institute/sktime/issues/805. Fixes https://github.com/alan-turing-institute/sktime/issues/1935. Fixes https://github.com/alan-turing-institute/sktime/issues/1435. Fixes https://github.com/alan-turing-institute/sktime/issues/1876.

#### What does this implement/fix? Explain your changes.

Convert `PeriodIndex` `absolute` to `DatetimeIndex`, as well as `cutoff` and `start` `Period` to `Timestamp` before taking the difference between the two.

This circumvents the following bug in pandas:
```
periods = pd.period_range(start="2021-01-01", periods=3, freq="2H")
periods - periods[0]
Out: Index([<0 * Hours>, <4 * Hours>, <8 * Hours>], dtype = 'object')
[v - periods[0] for v in periods]
Out: Index([<0 * Hours>, <2 * Hours>, <4 * Hours>], dtype='object')
```

Bug report scripts were added as explicit test cases and should be replaced later with lightweight unit tests.

#### Does your contribution introduce a new dependency? If yes, which one?

No.

##### For all contributions
- [X] I've added myself to the [list of contributors](https://github.com/alan-turing-institute/sktime/blob/main/.all-contributorsrc).
- [ ] Optionally, I've updated sktime's [CODEOWNERS](https://github.com/alan-turing-institute/sktime/blob/main/CODEOWNERS) to receive notifications about future changes to these files.
- [X] I've added unit tests and made sure they pass locally.

### CODE REVIEW

**Comment by fkiraly** [comment]
I think @eenticott-shell was already trying to fix this here: #1975 - assuming it is the same bug?
Perhaps it would be a good thing to check whether it's the same root cause, and, if yes, what broke in #1975 as @eenticott-shell tried to fix it.

**Comment by fkiraly** [comment]
uh-oh... are these the same failures as in #1975?

**Comment by eenticott-shell** [comment]
Thanks for continuing this, it looks like the tests are passing now.

File: sktime/forecasting/base/_fh.py [hunk 1]
```diff
@@ -273,23 +273,23 @@ def to_relative(self, cutoff=None):
...
 
-            if isinstance(absolute, pd.DatetimeIndex):
-                # We cannot use the freq from the the ForecastingHorizon itself (or its
-                # wrapped pd.DatetimeIndex) because it may be none for non-regular
-                # indices, so instead we use the freq of cutoff.
-                freq = _get_freq(cutoff)
+            # We cannot use the freq from the ForecastingHorizon itself (or its
+            # wrapped pd.DatetimeIndex) because it may be none for non-regular
+            # indices, so instead we use the freq of cutoff.
+            freq = _get_freq(cutoff)
```
Comment (fkiraly) [inline sktime/forecasting/base/_fh.py:279]: does cutoff always have a frequency?

File: sktime/forecasting/base/_fh.py [hunk 2]
```diff
@@ -273,23 +273,23 @@ def to_relative(self, cutoff=None):
...
 
+            if isinstance(absolute, pd.DatetimeIndex):
                 # coerce to pd.Period for reliable arithmetics and computations of
                 # time deltas
                 absolute = _coerce_to_period(absolute, freq)
                 cutoff = _coerce_to_period(cutoff, freq)
 
             # Compute relative values
-            relative = absolute - cutoff
+            relative = pd.Index([date - cutoff for date in absolute])
```
Comment (fkiraly) [inline sktime/forecasting/base/_fh.py:288]: can we avoid this back/forth conversion?

**Comment by khrapovs** [comment]
@fkiraly @eenticott-shell Finally, all tests pass! I have explicitly added all the bug report examples mentioned in the PR description as test cases. Obviously, at this state this PR is impossible to merge. It is too dirty and looks like a hack. I still can not comprehend the full scale of what is going on with those endless conversions in forecasting horizon. So, I will try to clean it up a bit. Any help is of course highly welcome!

File: sktime/forecasting/base/_fh.py (hunk 1 above)
Comment (khrapovs) [reply sktime/forecasting/base/_fh.py:279]: I wonder too. There are not enough tests to say for sure. Not all public methods are tested explicitly, but only implicitly through forecasters and transformers.

File: sktime/forecasting/base/_fh.py (hunk 2 above)
Comment (khrapovs) [reply sktime/forecasting/base/_fh.py:288]: this specific line is exactly the hack we need to circumvent the bug in pandas with difference between `PeriodIndex` and `Period` objects. converting the list to index is necessary as further down the road it is only index, but not a list is expected

File: sktime/forecasting/base/_fh.py (hunk 1 above)
Comment (fkiraly) [reply sktime/forecasting/base/_fh.py:279]: hm... somewhat troubling.
We may need an explicit design document on the different values that `cutoff` can (and is expected to) have; as well as behaviour of `ForecastingHorizon` around that. There is clearly an interaction with the `BaseForecaster`.

**Comment by fkiraly** [comment]
> Finally, all tests pass! I have explicitly added all the bug report examples mentioned in the PR description as test cases. Obviously, at this state this PR is impossible to merge. It is too dirty and looks like a hack. I still can not comprehend the full scale of what is going on with those endless conversions in forecasting horizon. So, I will try to clean it up a bit. Any help is of course highly welcome!

I would not mind merging this and cleaning up in a separate PR.
I've looked through it and think the flow and logic is preserved, it's not as bad as a hack as you may think.

Of course the flow and logic, and structure, is not perfect, and we should definitely revisit that, and architecture/design, when thinking about time stamp based horizons.

**Comment by khrapovs** [comment]
@fkiraly Ok, let's merge it then. I have cleaned up a bit the code, added more comments about pandas bug, a few TODO lines in the test module. I will come back to this either in https://github.com/alan-turing-institute/sktime/pull/1968, or separately.

File: sktime/forecasting/base/tests/test_fh.py [hunk 3]
```diff
@@ -264,3 +308,113 @@ def test_estimator_fh(freqstr):
...
+    forecaster.fit(y, fh=[1, 2, 3, 4, 5, 6])
+    y_pred = forecaster.predict()
+    pd.testing.assert_index_equal(
+        y_pred.index, pd.period_range("2019-01-02 00:00", periods=6, freq=freq)
+    )
+
+
+# TODO: Replace this long running test with fast unit test
+def test_theta_forecaster():
+    """Test example in notebook.
```
Comment (fkiraly) [inline sktime/forecasting/base/tests/test_fh.py:360]: the notebooks are already all tested as part of the CI/CD, i.e., that they all run without exception.

**Review by fkiraly — APPROVED** [review]
I think this is an improvement in cleaning up the code, and also a bugfix for the various bugs linked.

Since you worked on it too, @eenticott-shell, could you kindly have another look before we merge this?

File: sktime/forecasting/base/tests/test_fh.py (hunk 3 above)
Comment (khrapovs) [reply sktime/forecasting/base/tests/test_fh.py]: I have added this test only for my convenience. But now that everything is fixed, I have removed it.

**Comment by eenticott-shell** [comment]
Thank you for your work on this @khrapovs. The fix is much cleaner and tidier now. Given that is now passes all the tests hopefully we can be safe to assume that the coercing hasn't caused any other issues.

**Review by fkiraly — APPROVED** [review]
Looks good!
Let's just check whether all the tests pass with an update from `main`.

**Comment by fkiraly** [comment]
@eenticott-shell, @khrapovs, quick Q: did any of you report the `pandas` bug to `pandas` yet?
Let me know if not and/or whether you want to do it.

**Pull request merged**